zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ProcessInstanceMigrationClusteredTest.java

```diff
@@ -0,0 +1,201 @@
+
+  @Test
+  public void shouldRejectMigrationWhenMessageSubscriptionMigrationIsNotDistributed() {
+    // given
+    final DeploymentEvent deploymentEvent =
+        clientRule
+            .getClient()
+            .newDeployResourceCommand()
+            .addProcessModel(SOURCE_PROCESS, "process.bpmn")
+            .addProcessModel(TARGET_PROCESS, "targetProcess.bpmn")
+            .addProcessModel(TARGET_PROCESS_2, "targetProcess2.bpmn")
+            .send()
+            .join();
+    final var sourceProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(0).getProcessDefinitionKey();
+    final var targetProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(1).getProcessDefinitionKey();
+    final var secondTargetProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(2).getProcessDefinitionKey();
+
+    assertThat(
+            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.ACKNOWLEDGED)
+                .withDistributionPartitionId(2)
+                .findFirst())
+        .isPresent();
+    assertThat(
+            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.ACKNOWLEDGED)
+                .withDistributionPartitionId(3)
+                .findFirst())
+        .isPresent();
+
+    int processInstancePartitionId = 0;
+    int subscriptionPartitionId = 0;
+    long processInstanceKey = -1;
+
+    // To test multi partition behaviour of the migration, we need to ensure that the process
+    // instance and the message subscription are on different partitions
+    // TODO - make process instance partition id deterministic, so no need to loop
+    while (processInstancePartitionId == subscriptionPartitionId) {
+      final ProcessInstanceEvent processInstanceEvent =
+          clientRule
+              .getClient()
+              .newCreateInstanceCommand()
+              .processDefinitionKey(sourceProcessDefinitionKey)
+              .variables(Map.of("key1", "key1", "key2", "key2", "key3", "key3"))
+              .send()
+              .join();
+      processInstanceKey = processInstanceEvent.getProcessInstanceKey();
+
+      processInstancePartitionId = Protocol.decodePartitionId(processInstanceKey);
+      subscriptionPartitionId =
+          SubscriptionUtil.getSubscriptionPartitionId(BufferUtil.wrapString(MESSAGE_NAME), 3);
+    }
+
+    // Pause the stream processing on the partition leader for the message subscription
+    // Therefore, the message subscription migration distribution will not be processed
+    final var messageSubscriptionPartitionLeader =
+        clusteringRule.getLeaderForPartition(subscriptionPartitionId).getNodeId();
+    final var adminServiceMsgSubPartitionLeader =
+        clusteringRule
+            .getBroker(messageSubscriptionPartitionLeader)
+            .getBrokerContext()
+            .getBrokerAdminService();
+    adminServiceMsgSubPartitionLeader.pauseStreamProcessing();
```

Comment:
🔧 We can simplify this test case by making it an engine test and using the `ENGINE.writeRecords(...)` method to write two process instance migrations immediately after each other. In other words, the second migration happens before any events regarding the first have been appended, i.e. the second command is processed before any follow-up commands of the first command are processed, and before the other partition could acknowledge the distribution of the first migration.

The partition's logstream is sequential. So, we can focus our test case on the partition itself. We don't even have to pause the other partition. We control what the commands on the partition look like.

Reply:
@korthout This one will not test the same case entirely because we check if the message subscription is in pending state. If the second migration processed right after the first one, the subscription will not appear to be in pending state.